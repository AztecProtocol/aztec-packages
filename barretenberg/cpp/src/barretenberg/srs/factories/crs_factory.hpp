#pragma once
#include "barretenberg/common/mem.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/g2.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include <cstddef>

namespace bb::pairing {
struct miller_lines;
} // namespace bb::pairing

namespace bb::srs::factories {

/**
 * A prover crs representation.
 */
template <typename Curve> class ProverCrs {
  public:
    virtual ~ProverCrs() = default;
    ;

    /**
     *  @brief Returns the monomial points in a form to be consumed by scalar_multiplication pippenger algorithm.
     */
    virtual typename Curve::AffineElement* get_monomial_points() = 0;
    virtual size_t get_monomial_size() const = 0;
};

template <typename Curve> class VerifierCrs {
  public:
    virtual ~VerifierCrs() = default;
};
template <> class VerifierCrs<curve::BN254> {
    using Curve = curve::BN254;

  public:
    virtual Curve::G2AffineElement get_g2x() const = 0;
    /**
     * @brief As the G_2 element of the CRS is fixed, we can precompute the operations performed on it during the
     * pairing algorithm to optimize pairing computations.
     */
    virtual bb::pairing::miller_lines const* get_precomputed_g2_lines() const = 0;
    /**
     *  @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

template <> class VerifierCrs<curve::Grumpkin> {
    using Curve = curve::Grumpkin;

  public:
    /**
     * @brief Returns the G_1 elements in the CRS after the pippenger point table has been applied on them
     *
     */
    virtual Curve::AffineElement* get_monomial_points() const = 0;
    virtual size_t get_monomial_size() const = 0;
    /**
     * @brief Returns the first G_1 element from the CRS, used by the Shplonk verifier to compute the final
     * commtiment.
     */
    virtual Curve::AffineElement get_first_g1() const = 0;
};

/**
 * A factory class to return the prover crs and verifier crs on request.
 * You can construct an empty placeholder factory, because composers need to be given a factory at construction time.
 */
template <typename Curve> class CrsFactory {
  public:
    CrsFactory() = default;
    CrsFactory(CrsFactory&& other) = default;
    virtual ~CrsFactory() = default;
    virtual std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> get_prover_crs(size_t) { return nullptr; }
    virtual std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> get_verifier_crs([[maybe_unused]] size_t degree = 0)
    {
        return nullptr;
    }
};

} // namespace bb::srs::factories