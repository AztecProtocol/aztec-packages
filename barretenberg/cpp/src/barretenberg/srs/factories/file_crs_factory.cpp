#include "file_crs_factory.hpp"
#include "../io.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"
#include "barretenberg/ecc/curves/bn254/pairing.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include "barretenberg/ecc/scalar_multiplication/point_table.hpp"
#include "barretenberg/ecc/scalar_multiplication/scalar_multiplication.hpp"

namespace bb::srs::factories {

FileVerifierCrs<curve::BN254>::FileVerifierCrs(std::string const& path, const size_t)
    : precomputed_g2_lines((bb::pairing::miller_lines*)(aligned_alloc(64, sizeof(bb::pairing::miller_lines) * 2)))
{
    using Curve = curve::BN254;
    auto point_buf = scalar_multiplication::point_table_alloc<Curve::AffineElement>(1);
    srs::IO<Curve>::read_transcript_g1(point_buf.get(), 1, path);
    srs::IO<curve::BN254>::read_transcript_g2(g2_x, path);
    bb::pairing::precompute_miller_lines(bb::g2::one, precomputed_g2_lines[0]);
    bb::pairing::precompute_miller_lines(g2_x, precomputed_g2_lines[1]);
    first_g1 = point_buf[0];
}

FileVerifierCrs<curve::BN254>::~FileVerifierCrs()
{
    aligned_free(precomputed_g2_lines);
}

FileVerifierCrs<curve::Grumpkin>::FileVerifierCrs(std::string const& path, const size_t num_points)
    : num_points(num_points)
{
    using Curve = curve::Grumpkin;
    monomials_ = scalar_multiplication::point_table_alloc<Curve::AffineElement>(num_points);
    srs::IO<Curve>::read_transcript_g1(monomials_.get(), num_points, path);
    scalar_multiplication::generate_pippenger_point_table<Curve>(monomials_.get(), monomials_.get(), num_points);
    first_g1 = monomials_[0];
};

curve::Grumpkin::AffineElement* FileVerifierCrs<curve::Grumpkin>::get_monomial_points() const
{
    return monomials_.get();
}

size_t FileVerifierCrs<curve::Grumpkin>::get_monomial_size() const
{
    return num_points;
}

template <typename Curve>
FileCrsFactory<Curve>::FileCrsFactory(std::string path, size_t initial_degree)
    : path_(std::move(path))
    , degree_(initial_degree)
{}

template <typename Curve>
std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> FileCrsFactory<Curve>::get_prover_crs(size_t degree)
{
    if (degree != degree_ || !prover_crs_) {
        prover_crs_ = std::make_shared<FileProverCrs<Curve>>(degree, path_);
        degree_ = degree;
    }
    return prover_crs_;
}

template <typename Curve>
std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> FileCrsFactory<Curve>::get_verifier_crs(size_t degree)
{
    if (degree != degree_ || !verifier_crs_) {
        verifier_crs_ = std::make_shared<FileVerifierCrs<Curve>>(path_, degree);
        degree_ = degree;
    }
    return verifier_crs_;
}

template class FileProverCrs<curve::BN254>;
template class FileProverCrs<curve::Grumpkin>;
template class FileCrsFactory<curve::BN254>;
template class FileCrsFactory<curve::Grumpkin>;

} // namespace bb::srs::factories
