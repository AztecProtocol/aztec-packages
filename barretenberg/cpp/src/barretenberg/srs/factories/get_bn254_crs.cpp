#include "get_bn254_crs.hpp"
#include "barretenberg/api/exec_pipe.hpp"
#include "barretenberg/api/file_io.hpp"
#include "barretenberg/ecc/curves/bn254/g1.hpp"

namespace {
std::vector<uint8_t> download_bn254_g1_data(size_t num_points)
{
    size_t g1_end = num_points * sizeof(bb::g1::affine_element) - 1;

    std::string url = "https://crs.aztec.network/g1.dat";

    // IMPORTANT: this currently uses a shell, DO NOT let user-controlled strings here.
    std::string command = "curl -H \"Range: bytes=0-" + std::to_string(g1_end) + "\" '" + url + "'";

    auto data = bb::exec_pipe(command);
    // Header + num_points * sizeof point.
    if (data.size() < g1_end) {
        throw_or_abort("Failed to download g1 data.");
    }

    return data;
}

std::vector<uint8_t> download_bn254_g2_data()
{
    std::string url = "https://crs.aztec.network/g2.dat";
    // IMPORTANT: this currently uses a shell, DO NOT let user-controlled strings here.
    std::string command = "curl '" + url + "'";
    return bb::exec_pipe(command);
}
} // namespace

namespace bb {
std::vector<g1::affine_element> get_bn254_g1_data(const std::filesystem::path& path,
                                                  size_t num_points,
                                                  bool allow_download)
{
    // TODO(AD): per Charlie this should just download and replace the flat file portion atomically so we have no race
    // condition
    std::filesystem::create_directories(path);

    auto g1_path = path / "bn254_g1.dat";
    size_t g1_downloaded_points = get_file_size(g1_path) / sizeof(g1::affine_element);

    if (g1_downloaded_points >= num_points) {
        vinfo("using cached bn254 crs with num points ", std::to_string(g1_downloaded_points), " at ", g1_path);
        auto data = read_file(g1_path, num_points * sizeof(g1::affine_element));
        auto points = std::vector<g1::affine_element>(num_points);
        for (size_t i = 0; i < num_points; ++i) {
            points[i] = from_buffer<g1::affine_element>(data, i * sizeof(g1::affine_element));
        }
        return points;
    }

    if (!allow_download && g1_downloaded_points == 0) {
        throw_or_abort("bn254 g1 data not found and download not allowed in this context");
    } else if (!allow_download) {
        throw_or_abort(format("bn254 g1 data had ",
                              g1_downloaded_points,
                              " points and ",
                              num_points,
                              " were requested but download not allowed in this context"));
    }
    vinfo("downloading bn254 crs...");
    auto data = download_bn254_g1_data(num_points);
    write_file(g1_path, data);

    auto points = std::vector<g1::affine_element>(num_points);
    for (size_t i = 0; i < num_points; ++i) {
        points[i] = from_buffer<g1::affine_element>(data, i * sizeof(g1::affine_element));
    }
    return points;
}

g2::affine_element get_bn254_g2_data(const std::filesystem::path& path, bool allow_download)
{
    std::filesystem::create_directories(path);

    auto g2_path = path / "bn254_g2.dat";
    size_t g2_file_size = get_file_size(g2_path);

    if (g2_file_size == sizeof(g2::affine_element)) {
        auto data = read_file(g2_path);
        return from_buffer<g2::affine_element>(data.data());
    }
    if (!allow_download) {
        throw_or_abort("bn254 g2 data not found and download not allowed in this context");
    }
    auto data = download_bn254_g2_data();
    write_file(g2_path, data);
    return from_buffer<g2::affine_element>(data.data());
}
} // namespace bb
