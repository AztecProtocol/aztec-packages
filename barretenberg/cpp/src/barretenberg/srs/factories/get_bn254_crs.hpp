#pragma once
#include <barretenberg/ecc/curves/bn254/g1.hpp>
#include <barretenberg/ecc/curves/bn254/g2.hpp>
#include <filesystem>
#include <fstream>
#include <ios>

namespace bb {
std::vector<g1::affine_element> get_bn254_g1_data(const std::filesystem::path& path,
                                                  size_t num_points,
                                                  bool allow_download = true);
g2::affine_element get_bn254_g2_data(const std::filesystem::path& path, bool allow_download = true);
} // namespace bb
