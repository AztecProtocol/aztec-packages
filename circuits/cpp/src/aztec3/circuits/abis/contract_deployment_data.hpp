#pragma once
#include <barretenberg/crypto/generators/generator_data.hpp>
#include <barretenberg/stdlib/hash/pedersen/pedersen.hpp>
#include <barretenberg/stdlib/primitives/witness/witness.hpp>
#include <aztec3/utils/types/circuit_types.hpp>
#include <aztec3/utils/types/convert.hpp>
#include <aztec3/utils/types/native_types.hpp>

namespace aztec3::circuits::abis {

using aztec3::utils::types::CircuitTypes;
using aztec3::utils::types::NativeTypes;
using plonk::stdlib::witness_t;
using std::is_same;

template <typename NCT> struct ContractDeploymentData {
    typedef typename NCT::address address;
    typedef typename NCT::fr fr;
    typedef typename NCT::boolean boolean;

    fr constructor_vk_hash = 0;
    fr function_tree_root = 0;
    fr contract_address_salt = 0;
    address portal_contract_address = 0;

    boolean operator==(ContractDeploymentData<NCT> const& other) const
    {
        return constructor_vk_hash == other.constructor_vk_hash && function_tree_root == other.function_tree_root &&
               contract_address_salt == other.contract_address_salt &&
               portal_contract_address == other.portal_contract_address;
    };

    template <typename Composer>
    ContractDeploymentData<CircuitTypes<Composer>> to_circuit_type(Composer& composer) const
    {
        static_assert((std::is_same<NativeTypes, NCT>::value));

        // Capture the composer:
        auto to_ct = [&](auto& e) { return aztec3::utils::types::to_ct(composer, e); };

        ContractDeploymentData<CircuitTypes<Composer>> data = {
            to_ct(constructor_vk_hash),
            to_ct(function_tree_root),
            to_ct(contract_address_salt),
            to_ct(portal_contract_address),
        };

        return data;
    };

    template <typename Composer> ContractDeploymentData<NativeTypes> to_native_type() const
    {
        static_assert(std::is_same<CircuitTypes<Composer>, NCT>::value);
        auto to_nt = [&](auto& e) { return aztec3::utils::types::to_nt<Composer>(e); };

        ContractDeploymentData<NativeTypes> call_context = {
            to_nt(constructor_vk_hash),
            to_nt(function_tree_root),
            to_nt(contract_address_salt),
            to_nt(portal_contract_address),
        };

        return call_context;
    };

    template <typename Composer> void assert_is_zero()
    {
        static_assert((std::is_same<CircuitTypes<Composer>, NCT>::value));

        constructor_vk_hash.assert_is_zero();
        function_tree_root.assert_is_zero();
        contract_address_salt.assert_is_zero();
        portal_contract_address.to_field().assert_is_zero();
    }

    void set_public()
    {
        static_assert(!(std::is_same<NativeTypes, NCT>::value));

        constructor_vk_hash.set_public();
        function_tree_root.set_public();
        contract_address_salt.set_public();
        portal_contract_address.to_field().set_public();
    }

    fr hash() const
    {
        std::vector<fr> inputs = {
            constructor_vk_hash,
            function_tree_root,
            contract_address_salt,
            portal_contract_address.to_field(),
        };

        return NCT::compress(inputs, GeneratorIndex::CONTRACT_DEPLOYMENT_DATA);
    }
};

template <typename NCT> void read(uint8_t const*& it, ContractDeploymentData<NCT>& data)
{
    using serialize::read;

    read(it, data.constructor_vk_hash);
    read(it, data.function_tree_root);
    read(it, data.contract_address_salt);
    read(it, data.portal_contract_address);
};

template <typename NCT> void write(std::vector<uint8_t>& buf, ContractDeploymentData<NCT> const& data)
{
    using serialize::write;

    write(buf, data.constructor_vk_hash);
    write(buf, data.function_tree_root);
    write(buf, data.contract_address_salt);
    write(buf, data.portal_contract_address);
};

template <typename NCT> std::ostream& operator<<(std::ostream& os, ContractDeploymentData<NCT> const& data)
{
    return os << "constructor_vk_hash: " << data.constructor_vk_hash << "\n"
              << "function_tree_root: " << data.function_tree_root << "\n"
              << "contract_address_salt: " << data.contract_address_salt << "\n"
              << "portal_contract_address: " << data.portal_contract_address << "\n";
}

} // namespace aztec3::circuits::abis