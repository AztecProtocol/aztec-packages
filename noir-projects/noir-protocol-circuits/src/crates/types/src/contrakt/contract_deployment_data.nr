use crate::{
    address::EthAddress, contract_class_id::ContractClassId,
    constants::{CONTRACT_DEPLOYMENT_DATA_LENGTH, GENERATOR_INDEX__CONTRACT_DEPLOYMENT_DATA},
    grumpkin_point::GrumpkinPoint, hash::pedersen_hash, traits::{Deserialize, Hash, Serialize}
};

// docs:start:contract-deployment-data
struct ContractDeploymentData {
    public_key : GrumpkinPoint,
    initialization_hash : Field,
    contract_class_id : ContractClassId,
    contract_address_salt : Field,
    portal_contract_address : EthAddress,
}
// docs:end:contract-deployment-data

impl Eq for ContractDeploymentData {
    fn eq(self, other: Self) -> bool {
        self.public_key.eq(other.public_key) &
        self.initialization_hash.eq(other.initialization_hash) &
        self.contract_class_id.eq(other.contract_class_id) &
        self.contract_address_salt.eq(other.contract_address_salt) &
        self.portal_contract_address.eq(other.portal_contract_address)
    }
}

impl Hash for ContractDeploymentData {
    fn hash(self) -> Field {
        pedersen_hash(self.serialize(), GENERATOR_INDEX__CONTRACT_DEPLOYMENT_DATA)
    }
}

impl Serialize<CONTRACT_DEPLOYMENT_DATA_LENGTH> for ContractDeploymentData {
    fn serialize(self) -> [Field; CONTRACT_DEPLOYMENT_DATA_LENGTH] {
        [
            self.public_key.x,
            self.public_key.y,
            self.initialization_hash,
            self.contract_class_id.to_field(),
            self.contract_address_salt,
            self.portal_contract_address.to_field(),
        ]
    }
}

impl Deserialize<CONTRACT_DEPLOYMENT_DATA_LENGTH> for ContractDeploymentData {
    fn deserialize(serialized: [Field; CONTRACT_DEPLOYMENT_DATA_LENGTH]) -> Self {
        Self {
            public_key: GrumpkinPoint {
                x: serialized[0],
                y: serialized[1],
            },
            initialization_hash: serialized[2],
            contract_class_id: ContractClassId::from_field(serialized[3]),
            contract_address_salt: serialized[4],
            portal_contract_address: EthAddress::from_field(serialized[5]),
        }
    }
}

impl ContractDeploymentData {
    fn assert_is_zero(self) {
        self.public_key.assert_is_zero();
        assert(self.initialization_hash == 0);
        self.contract_class_id.assert_is_zero();
        assert(self.contract_address_salt == 0);
        self.portal_contract_address.assert_is_zero();
    }
}

#[test]
fn serialization_of_empty() {
    let data: ContractDeploymentData = dep::std::unsafe::zeroed();
    let serialized = data.serialize();
    let deserialized = ContractDeploymentData::deserialize(serialized);
    assert(data.eq(deserialized));
}

#[test]
fn empty_hash() {
    let data: ContractDeploymentData = dep::std::unsafe::zeroed();
    let hash = data.hash();

    // Value from contract_deployment_data.test.ts "computes empty item hash" test
    assert_eq(hash, 0x0e7babf59de8dfc7f5992cd34fb0066105d07ce67f68fc33d4a0a6a933a30405);
}
