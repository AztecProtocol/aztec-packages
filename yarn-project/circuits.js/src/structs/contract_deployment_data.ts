import { pedersenHash } from '@aztec/foundation/crypto';
import { EthAddress } from '@aztec/foundation/eth-address';
import { Fr, Point } from '@aztec/foundation/fields';
import { BufferReader, FieldReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { FieldsOf } from '@aztec/foundation/types';

import { CONTRACT_DEPLOYMENT_DATA_LENGTH, GeneratorIndex } from '../constants.gen.js';
import { PublicKey } from '../types/public_key.js';

/**
 * Contract deployment data in a TxContext
 * Not to be confused with NewContractData.
 */
export class ContractDeploymentData {
  constructor(
    /** Public key of the contract. */
    public publicKey: PublicKey,
    /** Hash of the initialization payload. */
    public initializationHash: Fr,
    /** Contract class identifier. */
    public contractClassId: Fr,
    /** Contract address salt (used when deriving a contract address). */
    public contractAddressSalt: Fr,
    /** Ethereum address of the portal contract on L1. */
    public portalContractAddress: EthAddress,
  ) {}

  static getFields(fields: FieldsOf<ContractDeploymentData>) {
    return [
      fields.publicKey,
      fields.initializationHash,
      fields.contractClassId,
      fields.contractAddressSalt,
      fields.portalContractAddress,
    ] as const;
  }

  toBuffer() {
    return serializeToBuffer(...ContractDeploymentData.getFields(this));
  }

  toFields(): Fr[] {
    const fields = serializeToFields(...ContractDeploymentData.getFields(this));
    if (fields.length !== CONTRACT_DEPLOYMENT_DATA_LENGTH) {
      throw new Error(
        `Invalid number of fields for ContractDeploymentData. Expected ${CONTRACT_DEPLOYMENT_DATA_LENGTH}, got ${fields.length}`,
      );
    }
    return fields;
  }

  /**
   * Returns an empty ContractDeploymentData.
   * @returns The empty ContractDeploymentData.
   */
  public static empty(): ContractDeploymentData {
    return new ContractDeploymentData(Point.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO, EthAddress.ZERO);
  }

  isEmpty() {
    return ContractDeploymentData.getFields(this).every(f => f.isZero());
  }

  /**
   * Deserializes contract deployment data rom a buffer or reader.
   * @param buffer - Buffer to read from.
   * @returns The deserialized ContractDeploymentData.
   */
  static fromBuffer(buffer: Buffer | BufferReader): ContractDeploymentData {
    const reader = BufferReader.asReader(buffer);
    return new ContractDeploymentData(
      reader.readObject(Point),
      Fr.fromBuffer(reader),
      Fr.fromBuffer(reader),
      Fr.fromBuffer(reader),
      reader.readObject(EthAddress),
    );
  }

  static fromFields(fields: Fr[] | FieldReader): ContractDeploymentData {
    const reader = FieldReader.asReader(fields);
    return new ContractDeploymentData(
      reader.readObject(Point),
      reader.readField(),
      reader.readField(),
      reader.readField(),
      reader.readObject(EthAddress),
    );
  }

  hash(): Fr {
    return pedersenHash(
      this.toFields().map(f => f.toBuffer()),
      GeneratorIndex.CONTRACT_DEPLOYMENT_DATA,
    );
  }
}
