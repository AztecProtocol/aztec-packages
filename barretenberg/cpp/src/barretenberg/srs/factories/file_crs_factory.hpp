#pragma once
#include "../io.hpp"
#include "barretenberg/ecc/curves/bn254/bn254.hpp"
#include "barretenberg/ecc/curves/grumpkin/grumpkin.hpp"
#include "barretenberg/ecc/scalar_multiplication/point_table.hpp"
#include "barretenberg/ecc/scalar_multiplication/scalar_multiplication.hpp"
#include "crs_factory.hpp"
#include <cstddef>
#include <utility>

namespace bb::srs::factories {

/**
 * Create reference strings given a path to a directory of transcript files.
 */
template <typename Curve> class FileCrsFactory : public CrsFactory<Curve> {
  public:
    FileCrsFactory(std::string path, size_t initial_degree = 0);
    FileCrsFactory(FileCrsFactory&& other) = default;

    std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> get_prover_crs(size_t degree) override;

    std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> get_verifier_crs(size_t degree = 0) override;

  private:
    std::string path_;
    size_t degree_;
    std::shared_ptr<bb::srs::factories::ProverCrs<Curve>> prover_crs_;
    std::shared_ptr<bb::srs::factories::VerifierCrs<Curve>> verifier_crs_;
};

template <typename Curve> class FileProverCrs : public ProverCrs<Curve> {
  public:
    FileProverCrs(const size_t num_points, std::string const& path)
        : num_points(num_points)
    {
        monomials_ = scalar_multiplication::point_table_alloc<typename Curve::AffineElement>(num_points);

        srs::IO<Curve>::read_transcript_g1(monomials_.get(), num_points, path);
        scalar_multiplication::generate_pippenger_point_table<Curve>(monomials_.get(), monomials_.get(), num_points);
    };

    typename Curve::AffineElement* get_monomial_points() { return monomials_.get(); }

    [[nodiscard]] size_t get_monomial_size() const { return num_points; }

  private:
    size_t num_points;
    std::shared_ptr<typename Curve::AffineElement[]> monomials_;
};

template <typename Curve> class FileVerifierCrs : public VerifierCrs<Curve> {
  public:
    FileVerifierCrs(std::string const& path, const size_t num_points);
};

template <> class FileVerifierCrs<curve::BN254> : public VerifierCrs<curve::BN254> {
    using Curve = curve::BN254;

  public:
    FileVerifierCrs(std::string const& path, const size_t num_points = 0);
    virtual ~FileVerifierCrs();
    Curve::G2AffineElement get_g2x() const override { return g2_x; };
    pairing::miller_lines const* get_precomputed_g2_lines() const override { return precomputed_g2_lines; };
    Curve::AffineElement get_first_g1() const override { return first_g1; };

  private:
    Curve::AffineElement first_g1;
    Curve::G2AffineElement g2_x;
    pairing::miller_lines* precomputed_g2_lines;
};

template <> class FileVerifierCrs<curve::Grumpkin> : public VerifierCrs<curve::Grumpkin> {
    using Curve = curve::Grumpkin;

  public:
    FileVerifierCrs(std::string const& path, const size_t num_points);
    virtual ~FileVerifierCrs() = default;
    Curve::AffineElement* get_monomial_points() const override;
    size_t get_monomial_size() const override;
    Curve::AffineElement get_first_g1() const override { return first_g1; };

  private:
    Curve::AffineElement first_g1;
    size_t num_points;
    std::shared_ptr<Curve::AffineElement[]> monomials_;
};

} // namespace bb::srs::factories
