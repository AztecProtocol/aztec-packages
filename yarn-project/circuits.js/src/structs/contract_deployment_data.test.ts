import { CONTRACT_DEPLOYMENT_DATA_LENGTH } from '../constants.gen.js';
import { makeContractDeploymentData } from '../tests/factories.js';
import { ContractDeploymentData } from './contract_deployment_data.js';

describe('ContractDeploymentData', () => {
  it(`serializes to buffer and deserializes it back`, () => {
    const expected = makeContractDeploymentData(1);
    const buffer = expected.toBuffer();
    const res = ContractDeploymentData.fromBuffer(buffer);
    expect(res).toEqual(expected);
    expect(res.isEmpty()).toBe(false);
  });

  it('number of fields matches constant', () => {
    const target = makeContractDeploymentData(327);
    const fields = target.toFields();
    expect(fields.length).toBe(CONTRACT_DEPLOYMENT_DATA_LENGTH);
  });

  it('computes empty hash', () => {
    const cd = ContractDeploymentData.empty();
    expect(cd.isEmpty()).toBe(true);

    const hash = cd.hash();
    expect(hash).toMatchSnapshot();

    // Value used in empty_hash test in contract_deployment_data.nr
    // console.log("hash", hash.toString());
  });
});
