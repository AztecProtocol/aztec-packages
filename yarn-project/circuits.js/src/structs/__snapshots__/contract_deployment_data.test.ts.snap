// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContractDeploymentData computes empty hash 1`] = `
Fr {
  "asBigInt": 6550888889575534006716962010714219565950255479539883359305705974567832585221n,
  "asBuffer": {
    "data": [
      14,
      123,
      171,
      245,
      157,
      232,
      223,
      199,
      245,
      153,
      44,
      211,
      79,
      176,
      6,
      97,
      5,
      208,
      124,
      230,
      127,
      104,
      252,
      51,
      212,
      160,
      166,
      169,
      51,
      163,
      4,
      5,
    ],
    "type": "Buffer",
  },
}
`;
